bgpd/bgp_route.c

```diff
@@ -10860,56 +10860,76 @@ struct peer_pcounts {
 	unsigned int count[PCOUNT_MAX];
 	const struct peer *peer;
 	const struct bgp_table *table;
+	safi_t safi;
 };
 
-static int bgp_peer_count_walker(struct thread *t)
+static void bgp_peer_count_proc(struct bgp_node *rn,
+				struct peer_pcounts *pc)
 {
-	struct bgp_node *rn;
-	struct peer_pcounts *pc = THREAD_ARG(t);
+	const struct bgp_adj_in *ain;
+	const struct bgp_path_info *pi;
 	const struct peer *peer = pc->peer;
 
-	for (rn = bgp_table_top(pc->table); rn; rn = bgp_route_next(rn)) {
-		struct bgp_adj_in *ain;
-		struct bgp_path_info *pi;
-
-		for (ain = rn->adj_in; ain; ain = ain->next)
-			if (ain->peer == peer)
-				pc->count[PCOUNT_ADJ_IN]++;
+	for (ain = rn->adj_in; ain; ain = ain->next)
+		if (ain->peer == peer)
+			pc->count[PCOUNT_ADJ_IN]++;
 
-		for (pi = bgp_node_get_bgp_path_info(rn); pi; pi = pi->next) {
+	for (pi = bgp_node_get_bgp_path_info(rn); pi; pi = pi->next) {
 
-			if (pi->peer != peer)
-				continue;
+		if (pi->peer != peer)
+			continue;
 
-			pc->count[PCOUNT_ALL]++;
+		pc->count[PCOUNT_ALL]++;
 
-			if (CHECK_FLAG(pi->flags, BGP_PATH_DAMPED))
-				pc->count[PCOUNT_DAMPED]++;
-			if (CHECK_FLAG(pi->flags, BGP_PATH_HISTORY))
-				pc->count[PCOUNT_HISTORY]++;
-			if (CHECK_FLAG(pi->flags, BGP_PATH_REMOVED))
-				pc->count[PCOUNT_REMOVED]++;
-			if (CHECK_FLAG(pi->flags, BGP_PATH_STALE))
-				pc->count[PCOUNT_STALE]++;
-			if (CHECK_FLAG(pi->flags, BGP_PATH_VALID))
-				pc->count[PCOUNT_VALID]++;
+		if (CHECK_FLAG(pi->flags, BGP_PATH_DAMPED))
+			pc->count[PCOUNT_DAMPED]++;
+		if (CHECK_FLAG(pi->flags, BGP_PATH_HISTORY))
+			pc->count[PCOUNT_HISTORY]++;
+		if (CHECK_FLAG(pi->flags, BGP_PATH_REMOVED))
+			pc->count[PCOUNT_REMOVED]++;
+		if (CHECK_FLAG(pi->flags, BGP_PATH_STALE))
+			pc->count[PCOUNT_STALE]++;
+		if (CHECK_FLAG(pi->flags, BGP_PATH_VALID))
+			pc->count[PCOUNT_VALID]++;
+		if (!CHECK_FLAG(pi->flags, BGP_PATH_UNUSEABLE))
+			pc->count[PCOUNT_PFCNT]++;
+
+		if (CHECK_FLAG(pi->flags, BGP_PATH_COUNTED)) {
+			pc->count[PCOUNT_COUNTED]++;
+			if (CHECK_FLAG(pi->flags, BGP_PATH_UNUSEABLE))
+				flog_err(
+					EC_LIB_DEVELOPMENT,
+					"Attempting to count but flags say it is unusable");
+		} else {
 			if (!CHECK_FLAG(pi->flags, BGP_PATH_UNUSEABLE))
-				pc->count[PCOUNT_PFCNT]++;
-
-			if (CHECK_FLAG(pi->flags, BGP_PATH_COUNTED)) {
-				pc->count[PCOUNT_COUNTED]++;
-				if (CHECK_FLAG(pi->flags, BGP_PATH_UNUSEABLE))
-					flog_err(
-						EC_LIB_DEVELOPMENT,
-						"Attempting to count but flags say it is unusable");
-			} else {
-				if (!CHECK_FLAG(pi->flags, BGP_PATH_UNUSEABLE))
-					flog_err(
-						EC_LIB_DEVELOPMENT,
-						"Not counted but flags say we should");
-			}
+				flog_err(
+					EC_LIB_DEVELOPMENT,
+					"Not counted but flags say we should");
 		}
 	}
+}
+
+static int bgp_peer_count_walker(struct thread *t)
+{
+	struct bgp_node *rn, *rm;
+	const struct bgp_table *table;
+	struct peer_pcounts *pc = THREAD_ARG(t);
+
+	if (pc->safi == SAFI_MPLS_VPN || pc->safi == SAFI_ENCAP
```

Comment:
can't we iterate over all AFI/SAFI with FOREACH_AFI_SAFI and just check for SAFI_MPLS_VPN/SAFI_ENCAP/SAFI_EVPN without adding safi_t into struct peer_pcounts?

Reply:
Valid point. However, bgp_peer_counts() receives afi and safi from the upper layer and bgp_peer_count_walker() receives the routing table pointer for that afi/safi (peer->bgp->rib[afi][safi]).So running the FOREACH_AFI_SAFI loop inside bgp_peer_count_walker(0 doesnot make much sense.

Reply:
I agree with @bisdhdh in that it does not make sense here to loop here over afi/safi. We just need the extra information here.
@ton31337: If you are worried about the size of struct peer_pcounts, I could reorder its fields to optimize alignment on x64. Also `.table` could be removed and replaced by an `afi_t` field, then the table could be extracted from `.peer` in `bgp_peer_count_walker`, which would again save 4 bytes on x64.

Reply:
@pogojotz I'm fine with this change, I just suggested without adding safi in addition.